airflow/contrib/hooks/gcp_vision_hook.py

```diff
@@ -0,0 +1,254 @@
+
+        if not product_id:
+            # Product id was generated by the API
+            product_id = self._get_autogenerated_id(response)
+            self.log.info('Extracted autogenerated Product ID from the response: %s', product_id)
+
+        return product_id
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def get_product(self, location, product_id, project_id=None, retry=None, timeout=None, metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductGetOperator`
+        """
+        client = self.get_client()
+        name = PRODUCT_NAME_TEMPLATE.format(project_id, location, product_id)
+        self.log.info('Retrieving Product: %s', name)
+        response = self._handle_request(lambda **kwargs: client.get_product(**kwargs), name=name, retry=retry,
+                                        timeout=timeout, metadata=metadata)
+        self.log.info('Product retrieved:\n%s', response)
+        return MessageToJson(response)
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def update_product(self, product, location=None, product_id=None, update_mask=None, project_id=None,
+                       retry=None, timeout=None, metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductUpdateOperator`
+        """
+        client = self.get_client()
+        product = self._determine_product_name(product, product_id, location, project_id)
+        self.log.info('Updating ProductSet: %s', product.name)
+        response = self._handle_request(lambda **kwargs: client.update_product(**kwargs), product=product,
+                                        update_mask=update_mask, retry=retry, timeout=timeout,
+                                        metadata=metadata)
+        self.log.info('Product updated:\n%s', response)
+        return MessageToJson(response)
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def delete_product(self, location, product_id, project_id=None, retry=None, timeout=None, metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductDeleteOperator`
+        """
+        client = self.get_client()
+        name = PRODUCT_NAME_TEMPLATE.format(project_id, location, product_id)
+        self.log.info('Deleting ProductSet: %s', name)
+        response = self._handle_request(lambda **kwargs: client.delete_product(**kwargs), name=name,
+                                        retry=retry, timeout=timeout, metadata=metadata)
+        self.log.info('Product with the name [%s] deleted:', name)
+        return response
+
+    def _handle_request(self, fun, **kwargs):
+        try:
+            return fun(**kwargs)
+        except GoogleAPICallError as e:
+            if isinstance(e, AlreadyExists):
```

Comment:
It looks good for me. @potiuk, @antonimaciej  what do you think?

Reply:
It's ok to use several except clauses, but I think the problem is that we are raising either Google API "AlreadyExist" or AirflowException from the same method. Another approach could be to have different exception deriving from AirflowException: for example ResourceAlreadyExist(AirflowException). This one should be thrown instead of google API AlreadyExist raised. This way we always have AirflowException from this method and it better separates concerns.

I also do not see much value in self.log.error() message as it is now :).

Reply:
In the original version (with `isinstance`) I think it is clearer that `AlreadyExists` inherits from `GoogleAPICallError`.

Reply:
We discussed before that wrapping an exception with AirflowException helps with getting the right stack trace (this error is not always ignored - sometimes it bubbles up). I think we agreed before that we wrap the original low-level exceptions with AirflowException in our hooks (rather than let it happen outside of "execute" method in airflow core) to get the right stack trace - this is what @antonimaciej was working on before. I think we should be consistent :)

I am not insisting on having this special exception but I am agains bubbling up the AlreadyExist error to the operator. Are we 100% sure "AlreadyExist" will always be handled separately and ignored in all current and future methods calling _handle_request?

I actually looked a bit more and I found related problem (another comment follows shortly).

Reply:
At the very least we should at least mention in the docs that AlreadyExists error might be returned and should be handled (in the docstrings of all methods using _handle_request that might return it).

Reply:
> Are we 100% sure "AlreadyExist" will always be handled separately and ignored in all current and future methods calling _handle_request?
That's why I like the verbosity of Java method signatures which list all checked exception that it can throw ;)

I looked at our previous operators from Q4 2018 and there's a lot of try/except clauses checking for exceptions different than `AirflowException`.

That said, I could handle the `AlreadyExists` error in the hook, however to me such business logic belongs in the operator.
In my opinion the hook should be "stupid", being the service layer only, and the operator should interpret its responses and behave accordingly if necessary. And that's what we've generally done in the previous operators.

**TL;DR: I see nothing bad in handling some exceptions in the operators.**